### PR TITLE
Show UAST from code

### LINKED ISSUE #48: Show UAST from code
The details need to worked out, but the code viewer (#47) should be able to provide the UAST from the code, using the `/parse` endpoint (#46).

### PR DESCRIPTION
**Pull request by smacker**
Part of #48 

screen shot 2018-06-20 at 16 25 13
screen shot 2018-06-20 at 16 25 24
screen shot 2018-06-20 at 16 25 43

Because this issue isn't about design there are few differences that will need to fix in sequential prs:
- show uast button is located in left pane not in the title of modal (it would complicate code)
- there is no top bar on UAST (with locations, server, query) yet. The PR is big enough already. Better implement separately.
- it doesn't return underlying error from bblfsh
- list of languages is hard-coded 
- there are no workarounds for java/bash/??? yet (languageMode != language for them)
- it works freaking slow

### CODE REVIEW

**Comment by bzz** [comment]
@smacker do you think the list should be moved as TODO items to another issue then?

**Comment by smacker** [comment]
I think it would be better to split TODO in separate issues:

1) show uast button is located in left pane not in the title of modal (it would complicate code)
2) there is no top bar on UAST (with locations, server, query) yet
3) bblfsh integration improvements
 - it doesn't return underlying error from bblfsh
 - list of languages is hard-coded
4) languages improvements
 - there are no workarounds for java/bash/??? yet (languageMode != language for them)
 - currently, we highlight only 6 languages which bblfsh knows. We can add more. Or not. 
5) frontend performance improvements

**Comment by smacker** [comment]
rebased

File: frontend/src/components/CodeViewer.js [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+    this.handleLangChange = this.handleLangChange.bind(this);
+    this.handleShowUastChange = this.handleShowUastChange.bind(this);
+    this.removeError = this.removeError.bind(this);
+  }
+
+  componentDidMount() {
+    api.detectLang(this.props.code).then(res => {
+      this.setState({ loading: false, language: res.language });
+    });
```
[the remark this answers is not in the document]
Comment (smacker) [reply frontend/src/components/CodeViewer.js]: fixed

File: frontend/src/components/CodeViewer.js [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+  handleShowUastChange() {
+    const showUast = !this.state.showUast;
+    if (showUast) {
+      this.parseCode();
+    }
+    this.setState({ showUast });
+  }
+
+  parseCode() {
+    this.setState({ error: null });
```
[the remark this answers is not in the document]
Comment (smacker) [reply frontend/src/components/CodeViewer.js]: fixed

**Comment by smacker** [comment]
> If the parsing fails, the uast shows an empty 'Node'. I think it would be better to show an empty div.

there is a bug in lib, it returns UAST for null. Issue: https://github.com/src-d/gitbase-playground/issues/126

> If you go back to 'Select language' the window should go back to the initial state, no right panel and uast checkbox disabled.

fixed

> The highlight looks like is not working properly. In some nodes it highlights the code in yellow, for some others an orange cursor appears. It also looks like it is off by one character.

that's exactly how it works in bblfsh dashboard. All the problems you mentioned aren't frontend related. They exist because of differences in drivers. One of the issues about it in dashboard: https://github.com/bblfsh/dashboard/issues/68

**Pull request merged**